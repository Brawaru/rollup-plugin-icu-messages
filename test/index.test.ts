import { fileURLToPath } from 'node:url'
import { dirname, resolve } from 'pathe'
import { rollup } from 'rollup'
import json from '@rollup/plugin-json'
import { describe, expect, it } from 'vitest'
import TOML from '@ltd/j-toml'
import { AnyMessage, icuMessages } from '..'

describe('plugin', () => {
  it('should generate bundle', async () => {
    const { generate } = await rollup({
      input: [
        resolve(
          dirname(fileURLToPath(import.meta.url)),
          'fixtures/normal/input.mjs',
        ),
      ],
      plugins: [icuMessages({ format: 'crowdin' })],
    })

    const { output } = await generate({
      format: 'esm',
    })

    expect(output).toHaveLength(1)
    expect(output[0]?.code).toMatchSnapshot()
  })

  it('should fail with unresolved formatter', async () => {
    let err: unknown
    try {
      await rollup({
        input: [
          resolve(
            dirname(fileURLToPath(import.meta.url)),
            'fixtures/normal/input.mjs',
          ),
        ],
        plugins: [icuMessages({ format: 'non-existent' })],
      })
    } catch (err_) {
      err = err_
    }

    expect(err).toHaveProperty('code', 'ROLLUP_ICU_FORMATTER_RESOLVE_ERROR')
  })

  it('should fail with pre-compiled JSON', async () => {
    let err: unknown

    try {
      const { generate } = await rollup({
        input: [
          resolve(
            dirname(fileURLToPath(import.meta.url)),
            'fixtures/normal/input.mjs',
          ),
        ],
        plugins: [json(), icuMessages({ format: 'crowdin' })],
      })

      await generate({
        format: 'esm',
      })
    } catch (err_) {
      err = err_
    }

    expect(err).toHaveProperty('pluginCode', 'ROLLUP_ICU_TRANSFORM_ERROR')
  })

  it('should prevent json plugins if specified', async () => {
    const { generate } = await rollup({
      input: [
        resolve(
          dirname(fileURLToPath(import.meta.url)),
          'fixtures/with-json/input.mjs',
        ),
      ],
      plugins: [
        json(),
        icuMessages({
          format: 'crowdin',
          experimental: { wrapJSONPlugins: true },
        }),
      ],
    })

    const { output } = await generate({
      format: 'esm',
    })

    expect(output).toHaveLength(1)
    expect(output[0]?.code).toMatchSnapshot()
  })

  it('respects parser options', async () => {
    const { generate } = await rollup({
      input: [
        resolve(
          dirname(fileURLToPath(import.meta.url)),
          'fixtures/tagless-parsing/input.mjs',
        ),
      ],
      plugins: [
        json(),
        icuMessages({
          format: 'crowdin',
          parserOptions: {
            [AnyMessage]: {
              ignoreTag: true,
            },
            'invalid-key': {
              ignoreTag: false,
            },
          },
          experimental: {
            wrapJSONPlugins: true,
          },
        }),
      ],
    })

    const { output } = await generate({
      format: 'esm',
    })

    expect(output).toHaveLength(1)
    expect(output[0]?.code).toMatchSnapshot()
  })

  it('should parse TOML when specified', async () => {
    const { generate } = await rollup({
      input: [
        resolve(
          dirname(fileURLToPath(import.meta.url)),
          'fixtures/toml/input.mjs',
        ),
      ],
      plugins: [
        icuMessages({
          include: '**/*.messages.toml',
          format: 'crowdin',
          parse(code) {
            return TOML.parse(code)
          },
        }),
      ],
    })

    const { output } = await generate({
      format: 'esm',
    })

    expect(output).toHaveLength(1)
    expect(output[0]?.code).toMatchSnapshot()
  })

  it('exposes filter in public API', () => {
    expect(icuMessages().api).toHaveProperty('filter')
  })

  // TODO: add more tests
})
