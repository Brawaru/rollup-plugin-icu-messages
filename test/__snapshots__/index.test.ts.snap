// Vitest Snapshot v1

exports[`plugin > should fail with pre-compiled JSON 1`] = `[Error: Cannot transform "d:\\Main\\Dev\\JavaScript\\nuxt-intl-workspace\\rollup-plugin-intl-messages\\test\\fixtures\\normal\\messages.json" due to SyntaxError: Unexpected token e in JSON at position 0. It appears that this file has already been transformed. Ensure that no other plugin parses this file. If you use \`@rollup/plugin-json\`, make sure to exclude through options all files that are transformed by this plugin.]`;

exports[`plugin > should fail with unresolved formatter 1`] = `[Error: Cannot resolve built-in formatter "non-existent". Valid formatters are: crowdin, default, index, lokalise, simple, smartling, transifex]`;

exports[`plugin > should generate bundle 1`] = `
"const greeting = [
	{
		type: 0,
		value: \\"Hello, \\"
	},
	{
		type: 1,
		value: \\"name\\"
	},
	{
		type: 0,
		value: \\"!\\"
	}
];
var messages = {
	greeting: greeting
};

function example() {
  // eslint-disable-next-line no-console
  console.log(messages);
}

export { example as default };
"
`;
